### PR TITLE
Add structural and behavior test files from Artemis

### PR DESCRIPTION
**Pull request by krusche**
This PR adds the files of the default Java programming exercise template.
Currently we copy lots of files into many different repositories that does not need to be change at all.
Putting those into a library such as artemis-java-test-sandbox makes much more sense, therefore I added this PR.
I will create a corresponding PR in Artemis which updates the Java programming exercise template based on these changes.

- [x] Deploy the snapshot version to Maven so that we can test it in an actual Artemis exercise
- [ ] Release a new official version of artemis-java-test-sandbox (e.g. 1.3.4) and use this in the Artemis template

Related Artemis PR https://github.com/ls1intum/Artemis/pull/2382

### CODE REVIEW

File: src/main/java/de/tum/in/test/api/structural/testutils/ClassNameScanner.java [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+     * This method recursively walks the actual folder file structure starting from the assignment folder and adds
+     * each type it finds e.g. filenames ending in .java to the passed JSON object.
+     * @param assignmentFolderName: The root folder where the method starts walking the project structure.
+     * @param node: The current node the method is visiting.
+     * @param foundTypes: The JSON object where the type names and packages get appended.
+     */
+    private void walkProjectFileStructure(String assignmentFolderName, File node, Map<String, List<String>> foundTypes) {
+        String fileName = node.getName();
+
+        if(fileName.contains(".java")) {
```
Comment (kloessst) [inline src/main/java/de/tum/in/test/api/structural/testutils/ClassNameScanner.java:226]: Won't work for Kotlin `.kv` files

File: src/main/java/de/tum/in/test/api/structural/testutils/ClassNameScanner.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            String[] fileNameComponents = fileName.split("\\.");
+
+            String className = fileNameComponents[fileNameComponents.length - 2];
+            String packageName = node.getPath().substring(0, node.getPath().indexOf(".java"));
```
Comment (kloessst) [inline src/main/java/de/tum/in/test/api/structural/testutils/ClassNameScanner.java:230]: Won't work for Kotlin .kv files

**Review by kloessst — CHANGES_REQUESTED** [review]
Tested this locally with the corresponding Artemis PR. Java template works as expected but Kotlin has problems with class scanning.

File: src/main/java/de/tum/in/test/api/structural/testutils/ClassNameScanner.java [hunk 3]
```diff
@@ -318,4 +336,42 @@ private void walkProjectFileStructure(String assignmentFolderName, File node,
...
+		 */
+		double distance = DAMERAU_LEVENSHTEIN.distance(a, b);
+		if (distance <= 1.0 && Math.max(a.length(), b.length()) > 2)
+			return true;
+		/**
+		 * We accept everything with a distance below three as typo. At three and above,
+		 * misspelled identifiers can be easily recognized by a human or might not be
+		 * spelling errors.
+		 */
+		if (distance > 2)
```
Comment (kloessst) [inline src/main/java/de/tum/in/test/api/structural/testutils/ClassNameScanner.java:375]: Is the comment false or is the distance literal incorrect? Souldn't it be "with a distance below **two** as typo"

File: src/main/java/de/tum/in/test/api/structural/AttributeTestProvider.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+	 * 
+	 * @return A dynamic test container containing the test for each class which is
+	 *         then executed by JUnit.
+	 * @throws URISyntaxException an exception if the URI of the class name cannot
+	 *                            be generated (which seems to be unlikely)
+	 */
+	protected DynamicContainer generateTestsForAllClasses() throws URISyntaxException {
+		List<DynamicNode> tests = new ArrayList<>();
+		if (structureOracleJSON == null) {
+			fail("The LocalAttributeTest test can only run if the structural oracle (test.json) is present. If you do not provide it, delete LocalAttributeTest.java!");
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/test/api/structural/AttributeTestProvider.java:47]: This needs to be adjusted now as well (same for the other classes).

File: src/main/java/de/tum/in/test/api/structural/AttributeTestProvider.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+				String expectedClassName = expectedClassPropertiesJSON.getString(JSON_PROPERTY_NAME);
+				String expectedPackageName = expectedClassPropertiesJSON.getString(JSON_PROPERTY_PACKAGE);
+				ExpectedClassStructure expectedClassStructure = new ExpectedClassStructure(expectedClassName,
+						expectedPackageName, expectedClassJSON);
+				tests.add(dynamicTest("testAttributes[" + expectedClassName + "]",
+						() -> testAttributes(expectedClassStructure)));
+			}
+		}
+		if (tests.isEmpty()) {
+			fail("No tests for attributes available in the structural oracle (test.json). Either provide attributes information or delete AttributeTest.java!");
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/test/api/structural/AttributeTestProvider.java:64]: This should still be correct?

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]

**Review by MaisiKoleni — APPROVED** [review]
Looks good to me.

**Review by kloessst — APPROVED** [review]
Skimmed through the code looks fine!

**Pull request merged**